### PR TITLE
[ingress-nginx] add prestop lifecycle hook for rbac-proxy

### PR DESCRIPTION
**Pull request by miklezzzz**
## Description

Add a lifecycle preStop hook with a `while` loop checking `controller's healthz 10254 port availability` to the `kube-rbac-proxy` container of an Ingress nginx's pod, preventing it from getting instantly not ready on termination.
Also, a minor change - a possibility to set `spec.waitLoadBalancerOnTerminating` to `0` value (previously, it would render to nil and wouldn't be accounted).
## Why do we need it, and what problem does it solve?

In case we use non-null value of the `--shutdown-grace-period` flag (`loadBalancer` and `hostPort` inlets) for ingress-nginx-controller, kube-rbac-proxy should continue running for the same time before exiting on shutdown. Otherwise we end up with a terminating and `not` ready pod as `kube-rbac-proxy` exits immediately and prevents liveness/readiness probes of the ingress-controller container from succeeding, rendering the container as not ready). Having a terminating and `NOT` ready pod doesn't allow k8s to apply such failover techniques as [ProxyTerminatingEndpoints](https://github.com/kubernetes/enhancements/blob/master/keps/sig-network/1669-proxy-terminating-endpoints/README.md) and their [alternatives](https://docs.cilium.io/en/stable/network/kubernetes/kubeproxy-free/#graceful-termination). It may result in a short downtime that starts as soon as the readiness probe fails 3 times (~6 seconds since deleting the pod) when the old pod is in terminating state (waiting for shutdown-grace-period), and a new one isn't come up yet.

## What is the expected result?

The `kube-rbac-proxy` container doesn't exit before the ingress-controller closes its healthz port, allowing liveness/readiness probes to succeed.

## Checklist
- [ ] The code is covered by unit tests.
- [x] e2e tests passed.
- [ ] Documentation updated according to the changes.
- [x] Changes were tested in the Kubernetes cluster manually.

## Changelog entries

```changes
section: ingress-nginx
type: chore
summary: Add a lifecycle prestop hook to kube-rbac-proxy.
impact: Nginx Ingress controller pods will be recreated consecutively.
impact_level: low
```

### CODE REVIEW

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @miklezzzz. I've started the workflow for label 'e2e/run/yandex-cloud'...

**Comment by apolovov** [comment]
Consider not using a sleep in preStop hook in favor of some check which detects the controller liveness.
Like we did in [istio sidecar](https://github.com/deckhouse/deckhouse/blob/3558d4fad1f0be3b56a00246ec5353c8dcb6aba8/modules/110-istio/templates/control-plane/iop.yaml#L212C22-L212C22).

File: modules/402-ingress-nginx/templates/controller/controller.yaml [hunk 1]
```diff
@@ -12,12 +12,18 @@
...
 {{- if $loadBalancer }}
   {{- $defaultGracePeriod = 120 }}
 {{- else if eq $crd.spec.inlet "HostWithFailover" }}
   {{- $defaultGracePeriod = 0 }}
 {{- end }}
+{{- if kindIs "invalid" $crd.spec.waitLoadBalancerOnTerminating }}
+  {{- $gracePeriod = $defaultGracePeriod }}
+{{- else }}
+  {{- $gracePeriod = $crd.spec.waitLoadBalancerOnTerminating | int }}
+{{- end }}
```
[the remark this answers is not in the document]
Comment (miklezzzz) [reply modules/402-ingress-nginx/templates/controller/controller.yaml:26]: Helm `default` regards int value of 0 as nil and if we set waitLoadBalancerOnTerminating to 0 in a CRD, its value is ignored and the `defaultGracePeriod` value is used. I thought to fix this behavior by using kindIs.

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @miklezzzz. I've started the workflow for label 'e2e/run/yandex-cloud'...

File: modules/402-ingress-nginx/templates/controller/controller.yaml [hunk 2]
```diff
@@ -329,6 +335,15 @@ spec:
                   resource: daemonsets
                   subresource: prometheus-protobuf-metrics
                   name: ingress-nginx
+        {{- if gt $gracePeriod 0 }}
```
Comment (apolovov) [inline modules/402-ingress-nginx/templates/controller/controller.yaml:338]: Do we need the if at all?

File: modules/402-ingress-nginx/templates/controller/controller.yaml (hunk 1 above)
[the remark this answers is not in the document]
Comment (alex123012) [reply modules/402-ingress-nginx/templates/controller/controller.yaml:26]: Maybe use `ternary` except of if-else?

[the remark this answers is not in the document]
Comment (miklezzzz) [reply modules/402-ingress-nginx/templates/controller/controller.yaml]: with `ternary` it's more compact, but less obvious, as for me. But I like it :-D
```
{{- $gracePeriod = (kindIs "invalid" $crd.spec.waitLoadBalancerOnTerminating) | ternary $defaultGracePeriod $crd.spec.waitLoadBalancerOnTerminating }}
```

File: modules/402-ingress-nginx/templates/controller/controller.yaml (hunk 2 above)
Comment (miklezzzz) [reply modules/402-ingress-nginx/templates/controller/controller.yaml]: I agree, it seems it makes sense to add that preStop hook unconditionally

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @miklezzzz. I've started the workflow for label 'e2e/run/yandex-cloud'...

**Review by yalosev — CHANGES_REQUESTED** [review]
Could you add a test please. In the templates_tests

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @miklezzzz. I've started the workflow for label 'e2e/run/yandex-cloud'...

**Comment by miklezzzz** [comment]
> Could you add a test please. In the templates_tests

added a couple of tests to check that `spec.waitLoadBalancerOnTerminating` is processed correctly.

**Review by yalosev — APPROVED** [review]

**Review by alex123012 — APPROVED** [review]

**Review by nabokihms — APPROVED** [review]

**Pull request merged**